src/Catalog/CommitCollector.cs

```diff
@@ -56,10 +57,25 @@ public CommitCollector(Uri index, Func<HttpMessageHandler> handlerFunc = null)
                         batch.CommitTimeStamp == lastBatch.CommitTimeStamp, 
                         cancellationToken);
 
-                    front.Value = batch.CommitTimeStamp;
-                    await front.Save(cancellationToken);
+                    // Only commit if this is the last batch that is going to be processed or if the commit timestamp
+                    // has changed from the previous batch. This is important because if two batches have the same
+                    // commit timestamp but processing the second fails, we should not progress the cursor forward.
+                    if (ReferenceEquals(batch, lastBatch))
+                    {
+                        front.Value = batch.CommitTimeStamp;
+                        await front.Save(cancellationToken);
+                        Trace.TraceInformation("CommitCatalog.Fetch front.Value saved due to last batch: {0}", front);
+                    }
+                    else if (previousCommitTimeStamp.HasValue && previousCommitTimeStamp != batch.CommitTimeStamp)
```

Comment:
> I'm more concerned with how jobs are now than what they need to be.

They are all designed to be idempotent and have demonstrated themselves to since their implementation. That being said, there are bugs, such as this one. And no doubt others that lead to all of the V3 inconsistencies. 

> In your last example a timestamp is insufficient to distinguish between completion of A or completion of A and B.

Correct. If there is an error in B, we should reprocess A. This happens all the time today in all of our jobs. When we are processing a single batch and a transient blob storage error occurs, we reprocess the whole commit.

Modifying the cursor to have more granularity (such as tracking commit timestamp as well as IDs within that commit timestamp that have been completed) is possible but won't help much in the 99% case where jobs just plod along happily without any transient errors.